crates/forge/src/link.rs

```diff
@@ -1,415 +1,203 @@
-                            contract: tc,
-                            known_contracts,
-                            id,
-                            extra,
-                            dependencies,
-                        };
-
-                        post_link(post_link_input)?;
-                        continue
-                    }
+    /// Performs DFS on the graph of link references, and populates `deps` with all found libraries.
+    fn collect_dependencies<'a>(
+        &'a self,
+        target: &'a ArtifactId,
+        deps: &mut BTreeSet<&'a ArtifactId>,
+    ) -> Result<()> {
+        let references = self
+            .contracts
+            .get(target)
+            .ok_or_eyre("target artifact must be present at given artifacts set")?
+            .all_link_references();
+        for (file, libs) in &references {
+            for contract in libs.keys() {
+                let id = self
+                    .find_artifact_id_by_library_path(file, contract, Some(&target.version))
+                    .ok_or_else(|| {
+                        eyre::eyre!(
+                            "wasn't able to find artifact for library {} at {}",
+                            file,
+                            contract
+                        )
+                    })?;
+                if deps.insert(id) {
+                    self.collect_dependencies(id, deps)?;
                 }
             }
+        }
 
-            rt.bytecode = Some(target_bytecode_runtime);
-            let tc = CompactContractBytecode {
-                abi: Some(abi.clone()),
-                bytecode: Some(target_bytecode),
-                deployed_bytecode: Some(rt),
-            };
-
-            let post_link_input =
-                PostLinkInput { contract: tc, known_contracts, id, extra, dependencies };
+        Ok(())
+    }
 
-            post_link(post_link_input)?;
+    /// Links given artifact with either given library addresses or address computed from sender and
+    /// nonce.
+    ///
+    /// Each key in `libraries` should either be a global path or relative to project root. All
+    /// remappings should be resolved.
+    ///
+    /// When calling for `target` being an external library itself, you should check that `target`
+    /// does not appear in `libs_to_deploy` to avoid deploying it twice. It may happen in cases
+    /// when there is a dependency cycle including `target`.
+    pub fn link_with_nonce_or_address<'a>(
+        &'a self,
+        libraries: Libraries,
+        sender: Address,
+        mut nonce: u64,
+        target: &'a ArtifactId,
+    ) -> Result<LinkOutput> {
+        // Library paths in `link_references` keys are always stripped, so we have to strip
+        // user-provided paths to be able to match them correctly.
+        let mut libraries = libraries.with_stripped_file_prefixes(self.root.as_path());
+
+        let mut needed_libraries = BTreeSet::new();
```

Comment:
At first I was wondering if this was safe, because you can have a library that itself is unlinked, so it's not deployable until all those unlinked references are resolved.

However at second thought, this is safe, because we know the addresses in such a scenario ahead of time, as long as it is deployed in the correct order.

I feel like we should add a note about that, wdyt?

Reply:
yep, it's safe because libraries doesn't use its dependencies during deployment

there is already a note about the deployment order, do you think it should be extended?
https://github.com/foundry-rs/foundry/blob/20aad6be1ce2c275d2510e0a00dfba50621b5949/crates/forge/src/link.rs#L28

Reply:
no it's prob fine